pkg/agentctl/sync.go

```diff
@@ -0,0 +1,122 @@
+package agentctl
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/grafana/agent/pkg/client"
+	"github.com/grafana/agent/pkg/prometheus/instance"
+	"gopkg.in/yaml.v2"
+)
+
+// ConfigSync loads YAML files from a directory and syncs them to the
+// provided PrometheusClient API. All YAML files will be synced and
+// must be valid.
+//
+// The base name of the YAML file (i.e., without the file extension)
+// is used as the config name.
+//
+// ConfigSync will completely overwrite the set of active configs
+// present in the provided PrometheusClient - configs present in the
+// API but not in the directory will be deleted.
+func ConfigSync(logger log.Logger, cli client.PrometheusClient, dir string) error {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	ctx := context.Background()
+	existing, err := cli.ListConfigs(ctx)
+	if err != nil {
+		return fmt.Errorf("could not list configs: %w", err)
+	}
+
+	cfgs, err := ConfigsFromDirectory(dir)
+	if err != nil {
+		return err
+	}
+
+	uploaded := map[string]struct{}{}
+
+	for _, cfg := range cfgs {
+		level.Info(logger).Log("msg", "uploading config", "name", cfg.Name)
+		err := cli.PutConfiguration(ctx, cfg.Name, cfg)
+		if err != nil {
+			return err
```

Comment:
I'm assuming this means partial errors are not supported. If we see an error we abort the sync, although some of it might already have happened - could we document this?

An alternative for the future might be to implement a dry-run endpoint/flag to validate before we execute.

Reply:
Completely agree, I'll create a issue for that after this is merged

Reply:
I've also removed the early return and replaced it with an error return just so we get an `os.Exit(1)` to let our tooling realize the upload failed.